File: Gravels Extended Battles 1.21/common/src/main/java/drai/dev/gravelsextendedbattles/showdown/ShowdownInjectionManager.java
package drai.dev.gravelsextendedbattles.showdown;

import java.nio.file.*;
import java.util.*;
import java.util.function.*;

public class ShowdownInjectionManager {

    public static void injectEntry(
            String targetFile,
            String targetText,
            String entryText
    ) {
        try {
            Path filePath = Path.of(targetFile);
            String content = Files.readString(filePath);

            // Find highest num
            java.util.regex.Pattern numPattern = java.util.regex.Pattern.compile("num:\\s*(\\d+)");
            java.util.regex.Matcher matcher = numPattern.matcher(content);

            StringBuilder builder = new StringBuilder();
            if (!builder.toString().contains(entryText)){
                builder.append(entryText);
            }

            int insertPos = content.lastIndexOf(targetText);
            if (insertPos == -1) {
                throw new RuntimeException("Can't find closing marker: " + targetText);
            }
            insertPos = insertPos + targetText.length();
            String newContent = content.substring(0, insertPos) + builder + content.substring(insertPos);
            Files.writeString(filePath, newContent);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: .gravels-extended-battles-new/common/src/main/kotlin/drai/dev/gravelsextendedbattles/showdown/ShowdownInjectionManager.kt
package drai.dev.gravelsextendedbattles.showdown

import java.nio.file.Files
import java.nio.file.Path
import java.util.regex.Pattern

object ShowdownInjectionManager {

    /**
     * Inserts [entryText] immediately after the last occurrence of [targetText].
     *
     * Does nothing if [entryText] is already present.
     */
    fun injectEntry(
        file: Path,
        targetText: String,
        entryText: String
    ) {
        var content = Files.readString(file)

        if (entryText in content) {
            return
        }

        val targetIndex = content.lastIndexOf(targetText)

        require(targetIndex >= 0) {
            "Can't find injection target '$targetText' in $file"
        }

        val insertIndex = targetIndex + targetText.length

        content = buildString(content.length + entryText.length) {
            append(content, 0, insertIndex)
            append(entryText)
            append(content, insertIndex, content.length)
        }

        Files.writeString(file, content)
    }

    /**
     * Replaces the first occurrence of [target] with [replacement].
     */
    fun replaceFirst(
        file: Path,
        target: Regex,
        replacement: String
    ) {
        val content = Files.readString(file)

        if (!target.containsMatchIn(content)) {
            return
        }

        val patched = target.replaceFirst(content, replacement)
        Files.writeString(file, patched)
    }
}
